### PR TITLE
[RLlib] RLTrainer is all you need.

### PR DESCRIPTION
**Pull request by kouroshHakha**
Signed-off-by: Kourosh Hakhamaneshi <kourosh@anyscale.com>

## Why are these changes needed?
The RLOptimizer seems to have become this shallow module that just adds to the complexity of the system. 
It basically is responsible for two things: 1) defining framework optimizers, 2) containing the loss logic. 

defining the optimizer inside this module creates a lot of un-necessary complexities when it comes to multi-agent RLOptimizers. By moving this logic to RLTrainer we can get rid of these complexities. 

Since compute_loss is also a state-less function we can easily move that to RLTrainer as well. Now all users have to do is to extend RLTrainer directly to customize the training phase of their algorithm. For example BCOptimizer will now become part of BCRLTrainer's implementation where all I have to do is to optionally override the `_configure_optimizer()` and write `compute_loss`. compute_loss will be written as a multi-agent loss. This is where the first-class-ness of MARL comes into play. It will make very complicated MARL communication patterns possible and also extremely easy. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/rl_trainer/rl_trainer.py [hunk 1]
```diff
@@ -217,39 +280,49 @@ def remove_module(self, module_id: ModuleID) -> None:
 
         """
         self._module.remove_module(module_id)
-        self._rl_optimizer.remove_optimizer(module_id)
 
-    def init_trainer(self) -> None:
+        # rerun make_optimizers to update the params and optimizer
+        self.make_optimizers()
+
+    def make_module(self) -> RLModule:
```
Comment (avnishn) [inline rllib/core/rl_trainer/rl_trainer.py:287]: returns a MultiAgentRLModule

**Review by avnishn — APPROVED** [review]
lgtm comments and type hints being fixed

File: rllib/core/optim/tests/test_rl_optimizer_tf.py [hunk 2]
```diff
@@ -5,19 +5,13 @@
 import unittest
 
 import ray
```
Comment (kouroshHakha) [inline rllib/core/optim/tests/test_rl_optimizer_tf.py:7]: You can ignore everything that is under optim. Since these tests are removed from CI anyway.

**Comment by gjoliver** [comment]
Test failures do not seem relevant

**Pull request merged**